File: lc_cpp/lcCpp/math/perfectPairs.cpp
//
// Created by 24276 on 2025/8/17.
//
typedef long long ll;
#include <vector>
using namespace std;
class Solution {
public:
    long long perfectPairs(vector<int>& nums) {
        ll ans = 0;
        int left = 0;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            nums[i] = abs(nums[i]);
        }
        sort(nums.begin(), nums.end());
        for (int i = 0; i < n; i++) {
            while (nums[left] * 2 < nums[i]) {
                left++;
            }
            ans += i - left;
        }
        return ans;
    }
};

File: lc_py/lcPy/math/perfectPairs.py
from typing import List


class Solution:
    def perfectPairs(self, nums: List[int]) -> int:

        for i, x in enumerate(nums):
            nums[i] = abs(x)

        nums.sort()
        ans = left = 0

        for j, b in enumerate(nums):
            while nums[left] * 2 < abs(b):
                left += 1
            ans += j - left

        return ans


if __name__ == '__main__':
    s = Solution()
    s.perfectPairs([0, 1, 2, 3])